src/skottie-player.ts

```diff
@@ -554,14 +595,484 @@ export class SkottiePlayer extends LitElement {
     const file = files?.[0];
     if (file && mapKey) {
       const reader = new FileReader();
-      reader.readAsArrayBuffer(file);
-      reader.onload = () => {
-        (this.assets as any)[mapKey] = reader.result;
-        this.updatePlayer();
+      if (mapKey === 'Image_0.jpeg') {
+        this.updateColors(reader, file, mapKey);
+      } else {
+        this.readAssetFile(reader, file, mapKey);
       }
     }
   }
 
+  /**
+   * Updates template colors based on an uploaded image. Also reads in the
+   * uploaded image as an image asset.
+   * 
+   * @param reader  Used when reading image as asset
+   * @param file    Image file
+   * @param mapKey  Name of asset, should always be "Image_0.jpeg"
+   */
+  private updateColors(reader: FileReader, file: File, mapKey: string): void {
+    const url = URL.createObjectURL(file);
+    const oldCanvas = document.getElementById('canvas');
+    if (oldCanvas) {
+      oldCanvas.remove();
+    }
+    const img = new Image();
+    img.onload = () => { 
+      const canvasElement = document.createElement('canvas');
+      canvasElement.setAttribute('id', 'canvas');
+      canvasElement.setAttribute('width', this.width.toString());
+      canvasElement.setAttribute('height', this.height.toString());
+      const context = canvasElement.getContext('2d');
+      if (context) {
+        context.drawImage(img, 0, 0, img.width * this.height / img.height, this.height);
+        const myData = context.getImageData(0, 0, img.width * this.height / img.height, this.height);
+        
+        const k = 5;
+        let [centers, centerCounts] = this.getDominantColors(myData, k);
+        let [index, max] = [0, 0];
+        for (let i = 0; i < centerCounts.length; i++) {
+          if (centerCounts[i] > max) {
+            max = centerCounts[i];
+            index = i;
+          }
+        }
+        const color = centers[index];
+        console.log(color);
+        console.log(centers);
+
+        this.setTemplateColors(color);
+        this.updateJson();
+        this.readAssetFile(reader, file, mapKey);
+      }
+      URL.revokeObjectURL(url);
+    }
+    img.src = url;
+  }
+
+  /**
+   * Takes in an ImageData object and finds the k most dominant colors from
+   * the image data. Also returns the number of pixels assoiated with each 
+   * color's cluster.
+   * @param myData  ImageData generated from a HTML Canvas element
+   */
+  private getDominantColors(myData: ImageData, k: number): Array<any> {
```

Comment:
Could we move getDominantColors() into its own .ts file for better readability? A common source code management practice is to separate utility functions (such as getDominantColors) from application code (such as skottie-player.js). This can help with code maintenance, and also improves readability (i.e. avoid creating a gigantic file).